Docs: FastStore API and API extensions (#1969)
vtex/faststore: apps/site/pages/docs/api-extensions/extending-api-schema.mdx

---
title: '  Extending API schemas'
sidebar_label: 'API Extensions'
---

import { Callout, Step, Steps, Tabs, Tab } from 'nextra-theme-docs'

<header>

# Extending API schemas 

</header>

In this guide, learn how to extend VTEX and third-party APIs to customize them according to your store's needs, allowing you to efficiently retrieve and manipulate the necessary data.

---

## Before you start

> <Steps>

Author: Remove the Steps component.

### Avoid over-fetching data on pages
 
Even though you can add information to the FastStore API schema, you must be careful not to over-fetch data on your pages. See the [best practices for fetching data on your storefront](/docs/api-extensions#best-practices-for-fetching-data).
 
</Steps>

---

## Extending VTEX API Schemas 

FastStore streamlines the way of using data from VTEX APIs that the FastStore API does not expose.

As FastStore uses GraphQL, it's required to write [type definitions](https://graphql.org/learn/schema/#object-types-and-fields) and [resolvers](https://graphql.org/learn/execution/#root-fields-resolvers) to fetch the data you need. 
To define new type definitions and resolvers for FastStore using GraphQL, learn how to create a specific directory structure in the following step by step. 

---

### Step by step
#### Step 1 - Preparing the folders and files 

To declare a new type definition and resolver you need to create a directory structure:

1. In you store repository, go to the `src` folder and create the `graphql` subfolder. You can do this by running the following command in your terminal:

> <Tabs items={['macOs and Linux', 'Windows']}>

Author: The same command here. We can remove the Tabs component.

---
title: '  Extending API schemas'
sidebar_label: 'API Extensions'
---

import { Callout, Step, Steps, FileTree, Tabs, Tab } from 'nextra-theme-docs'
import { ExtendableQueryTable } from './../../../components/ExtendableQueryTable'

<header>

# Extending API schemas 

</header>

In this guide, learn how to extend VTEX and third-party APIs to customize them according to your store's needs, allowing you to efficiently retrieve and manipulate the necessary data.

---

## Before you start

<Steps>
### Avoid over-fetching data on pages
 
Even though you can add information to the FastStore API schema, you must be careful not to over-fetch data on your pages. See the [best practices for fetching data on your storefront](/docs/api-extensions#best-practices-for-fetching-data).
 
### Ensure your development server is active
 

> Ensure that your development server is up and running while developing to see the changes you make. You can do this by running the command `yarn dev` in your terminal, which will monitor the changes.

Reviewer: This is not necessary.

To define new type definitions and resolvers for FastStore using GraphQL, learn how to create a specific directory structure in the following step by step. 

---

### Step by step
#### Step 1 - Preparing the folders and files 

To declare a new type definition and resolver you need to create a directory structure:

1. In you store repository, go to the `src` folder and create the `graphql` subfolder. You can do this by running the following command in your terminal:

<Tabs items={['macOs and Linux', 'Windows']}>
  <Tab>
  ```bash copy
  mkdir graphql
  ```
  </Tab>

  <Tab>
   ```bash copy
  mkdir graphql
  ```
  </Tab>
</Tabs>

2. Inside the new `graphql` folder, create the `vtex` subfolder.

<Tabs items={['macOs and Linux', 'Windows']}>
  <Tab>
  ```bash copy
  mkdir vtex
  ```
  </Tab>

  <Tab>
   ```bash copy
  mkdir vtex
  ```
  </Tab>
</Tabs>

3. In `vtex`, create `resolvers` and `typeDefs` subfolders.
<Tabs items={['macOs and Linux', 'Windows']}>
  <Tab>
  ```bash copy
  mkdir resolvers/typeDefs
  ```
  </Tab>

  <Tab>
   ```bash copy
  mkdir resolvers\typeDefs
  ```
  </Tab>
</Tabs>

<Callout type="warning" emoji="⚠️">

>   The `src/graphql/vtex/resolvers` and `src/graphql/vtex/typeDefs` path is mandatory, but you can organize the files inside those directory as you wish.

Reviewer: ```suggestion
  The `src/graphql/vtex/resolvers` and `src/graphql/vtex/typeDefs` path are mandatory, but you can organize the files inside those directories as you wish.
```

Are they? You can have a store without these paths. In what sense are they mandatory? I also think you can have `src/graphql/vtex/resolvers` and don't have `src/graphql/vtex/typeDefs`, but @eduardoformiga can confirm this.

</Callout>

4. Create a `index.ts` file inside the `resolvers` folder.

Once you have created the essentials folders, you will have a folder structure for the VTEX API Schema Extensions in the `src/graphql/vtex` folder similiar to the the following:

```ts copy
starter.store/
└─ src/
    └─ graphql/
        └─ vtex/
          ├─ resolvers/
              └─ index.ts
              └─ <resolverName>.ts
              └─ …
          ├─ typeDefs/
              └─<typeName>.graphql
              └─ …
```

<Callout type="info" emoji="ℹ️">
  Note that in the code example above, the type definitions and resolvers file were are created ad `<resolverName>.ts` and `<typeName>.graphql`.
</Callout>

> #### Step 2 - Creating type definitions (`TypeDefs`)

Reviewer: ```suggestion
#### Step 2 - Creating type definitions (`typeDefs`)
```

I'd change casing to avoid confusion. Using this casing in the folder name makes it don't work.

```ts copy filename="graphql/vtex/resolvers/product.ts"

import type { StoreProductRoot } from '@faststore/api'

const productResolver = {
  StoreProduct: {
    customData: (root: StoreProductRoot) => {
      return 'My item id: ' + root.itemId
    },
  },
}

export default productResolver
```

Note that you can leverage the use of TypeScript by typing the root param as [StoreProductRoot](https://v1.faststore.dev/reference/api/root-object-fields#root-product):

![using-root-field-type](https://vtexhelp.vtexassets.com/assets/docs/src/root-field-ts___3a10ec8a1aeba9770c7ee982995d55bc.gif)

3. Open the `index.ts` file inside the `vtex/resolvers` folder and import the resolver, `StoreProductResolver`, you created in the previous step. 

4. Create a `resolvers` object that includes the imported resolver(s).

```ts copy filename="graphql/vtex/resolvers/index.ts"

import { default as StoreProductResolver } from './product'

const resolvers = {
  ...StoreProductResolver,
}

export default resolvers
```

Once you have defined these files in the `vtex` folder, the new fields are already available to use. 

<Callout type="warning" emoji="⚠️">

>   If the changes you made still not available, refer to the troubleshooting section [GraphQL changes not visible during development](/docs/api-extension/troubleshooting#graphql-changes-not-visible-during-development).

Reviewer: ```suggestion
  If the changes you made are still not available, refer to the troubleshooting section [GraphQL changes not visible during development](/docs/api-extension/troubleshooting#graphql-changes-not-visible-during-development).
```

</Callout>
 

> ## Extending third-party API schemas

Reviewer: ```suggestion
## Extending FastStore API with third-party API schemas
```

The third-party API schemas don't exist until the user defines it. We are actually extending FastStore API (in this case, always adding, not modifying its behavior) with third-party API schemas.

As stores tend to grow, so does the possibility of consuming new data that is not provided by default in the FastStore API or other VTEX APIs. 
As such, FastStore must consume new data from third-party APIs.

### Step by step
#### Step 1- Preparing the folders and files

<Callout type="warning" emoji="⚠️">

>   If you already have a directory structure created in the [Extending VTEX API Schemas](/docs/api-extension/extending--api-schema#step-1---preparing-the-folders-and-files),
>   for third-party API schemas you only need to add the `thirdParty` folder inside `src/graphql`.

Reviewer: ```suggestion
  For third-party API schemas you only need to add the `thirdParty` folder inside `src/graphql`.
```

I'd re-frase it so users know they shouldn't use the Extending VTEX API Schemas directory, but I wouldn't put it as a condition. They'll need the `thirdParty` folder anyways.

</Tabs>

2. Inside the new `graphql` folder, create the `thirdParty` subfolder.

<Tabs items={['macOs and Linux', 'Windows']}>
  <Tab>
  ```bash copy
  mkdir thirdParty
  ```
  </Tab>

  <Tab>
   ```bash copy
  mkdir thirdParty
  ```
  </Tab>
</Tabs>

3. In `thirdParty`, create `resolvers` and `typeDefs` subfolders.
<Tabs items={['macOs and Linux', 'Windows']}>
  <Tab>
  ```bash copy
  mkdir resolvers/typeDefs
  ```
  </Tab>

  <Tab>
   ```bash copy
  mkdir resolvers\typeDefs
  ```
  </Tab>
</Tabs>

<Callout type="warning" emoji="⚠️">
  The `src/graphql/vtex/resolvers` and `src/graphql/vtex/typeDefs` path is mandatory, but you can organize the files inside those directory as you wish.
</Callout>

4. Create a `index.ts` file inside the `resolvers` folder.

Once you have created the essentials folders, you will have a folder structure similiar to the the following:

```ts copy
starter.store/
└── src/
    └── graphql/
        ├── thirdParty/
        │   ├── resolvers/
        │       └── index.ts
        │       └── <resolverName>.ts
        │       └── …
        │   ├── typeDefs/
        │       └── <typeName>.graphql
        │       └── …
```

<Callout type="info" emoji="ℹ️">

>   This directory structure distinguishes between data 
>   originating from VTEX and data from third-party APIs, which helps in debugging.

Reviewer: ```suggestion
  This directory structure distinguishes between data 
  originating from VTEX and data from third-party APIs.
```

I'd remove this part or complement it further. I don't think developers will find this argument compelling enough, so I'd rather not have it.

</Callout>

#### Step 2- Creating type definitions and resolvers

Following this structure, you can use the same steps mentioned in Extending VTEX API schemas, [Step 2 - Creating type definitions (`TypeDefs`)](/docs/api-extension/extending-api-schema#step-2---creating-type-definitions-typedefs) and [Step 3 - Creating resolvers](/docs/api-extension/extending-api-schema#step-3---creating-resolvers).
However, use the `thirdParty` folder, placing resolvers and typeDefs inside their respective folders and importing resolvers into the `index.ts` file.

#### Step 3- Creating query or mutation schemas

> When developing a third-party extension, it's essential to create a Query or Mutation schema. This schema is crucial for generating new query entry points to access data. 

Reviewer 2: I'd rephrase this
```suggestion
It's important to mention that it's possible to extend the existing [FastStore API Objects](https://v1.faststore.dev/reference/api/objects) as mentioned in the [Extending VTEX API Schemas](/docs/api-extensions/extending-api-schema#extending-vtex-api-schemas), or even create new queries entry points.
```

> Below, you'll find an example of code for implementing a third-party extension.

Reviewer 2: ```suggestion
Below, you'll find an example of code for implementing a third-party extension using new queries.
```

> 1. Create the `extra.graphql` file inside the `graphql/thirdParty/typeDefs` folder. The `extra.graphql` defines the type definitions in in a GraphQL schema file.

Reviewer 2: ```suggestion
1. Create the `<typeName>.graphql` file inside the `graphql/thirdParty/typeDefs` folder. As we are creating new queries, the name of the file that defines the type definitions can be `extra.graphql`.
```

```ts copy filename="graphql/thirdParty/typeDefs/extra.graphql"

type ExtraData {
  """
  Data customizing ExtraData
  """
  data: String!
}

type Query {
  extraData: ExtraData
  namedExtraData(name: String!): ExtraData
}
```

> 2. Create the `extra.ts` file inside the the `graphql/thirdParty/resolvers` folder. This file defines the resolvers that handle the logic for the defined queries or mutations. 

Reviewer 2: ```suggestion
2. Create a new `<resolver>.ts` file inside the `graphql/thirdParty/resolvers` folder. Let's continue the example, so the file name can be `extra.ts`. This file defines the resolvers that handle the logic for the defined queries or mutations.
```

```ts copy filename="graphql/thirdParty/resolvers/extra.ts"

const extraDataResolver = {
  Query: {
    extraData: () => {
      return {
        data: 'Extra data',
      }
    },
    namedExtraData: (_, { name }) => {
      return {
        data: `Named extra data: ${name}`,
      }
    },
  },
}

export default extraDataResolver
```
3 - In the `graphql/thirdParty/resolvers` folder, create an `index.ts` file to combine all the resolvers into a single object.

```ts copy filename="graphql/thirdParty/resolvers/index.ts"

import { default as StoreExtraResolver } from './extra'

const resolvers = {
  ...StoreExtraResolver,
}

export default resolvers
```

4 - With the schema and resolvers in place, you can query the data using GraphQL queries. Here's an example query to access the extra data:

```ts copy
query {
	extraData {
		data
	}
	namedExtraData(name: "Hello") {
		data
	}
}
```

Once you have defined these files in the `thirdParty` folder, you can query the data you defined. 

<Callout type="warning" emoji="⚠️">

>   If the changes you made still not available, refer to the troubleshooting section [GraphQL changes not visible during development](/docs/api-extension/troubleshooting#graphql-changes-not-visible-during-development).

Reviewer 2: ```suggestion
  If the changes you made are still not available, refer to the troubleshooting section [GraphQL changes not visible during development](/docs/api-extension/troubleshooting#graphql-changes-not-visible-during-development).
```
